Dockerfile

```diff
@@ -23,4 +25,4 @@ COPY app /app
 
 EXPOSE 8000
 
-CMD ["python", "-m", "flask", "run"]
+ENTRYPOINT ["python", "run.py"]
```

Comment:
I tried to use ENTRYPOINT but  if you need to override it's better to use CMD.

Reply:
I also tried both and only got a graceful exit with `ENTRYPOINT`. I'm not sure why; it could have been something else, but I'm about 75% sure that reverting back to using `ENTRYPOINT` seemed to help.

Reply:
It coud be that `CMD` needs `"run.py"` instead of `"run"`, but I'm not sure.

Reply:
```sh
Step 12/12 : CMD ["python", "flask", "-m", "run.py"]
 ---> Running in a994d6d7a460
Removing intermediate container a994d6d7a460
 ---> 541ac4b41f7f

Successfully built 541ac4b41f7f
Successfully tagged resources:latest
Starting pg01 ... done
Recreating resources01 ... done
Attaching to pg01, resources01
pg01         | 2018-10-25 03:56:28.006 UTC [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
pg01         | 2018-10-25 03:56:28.006 UTC [1] LOG:  listening on IPv6 address "::", port 5432
pg01         | 2018-10-25 03:56:28.010 UTC [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
pg01         | 2018-10-25 03:56:28.030 UTC [19] LOG:  database system was shut down at 2018-10-25 03:56:12 UTC
pg01         | 2018-10-25 03:56:28.035 UTC [1] LOG:  database system is ready to accept connections
resources01  | (null): can't open file 'flask': [Errno 2] No such file or directory
resources01 exited with code 2
```

Reply:
```sh
Step 12/12 : CMD ["python", "-m", "flask", "run"]
 ---> Running in 999fb985a7f6
Removing intermediate container 999fb985a7f6
 ---> 799def31e222

Successfully built 799def31e222
Successfully tagged resources:latest
Starting pg01 ... done
Recreating resources01 ... done
Attaching to pg01, resources01
pg01         | 2018-10-25 03:58:30.034 UTC [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
pg01         | 2018-10-25 03:58:30.034 UTC [1] LOG:  listening on IPv6 address "::", port 5432
pg01         | 2018-10-25 03:58:30.039 UTC [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
pg01         | 2018-10-25 03:58:30.057 UTC [20] LOG:  database system was shut down at 2018-10-25 03:58:17 UTC
pg01         | 2018-10-25 03:58:30.062 UTC [1] LOG:  database system is ready to accept connections
resources01  |  * Serving Flask app "run.py"
resources01  |  * Environment: production
resources01  |    WARNING: Do not use the development server in a production environment.
resources01  |    Use a production WSGI server instead.
resources01  |  * Debug mode: off
resources01  |  * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
```
The command `make` doesn't exit. The process stays open.

Reply:
```sh
Step 12/12 : ENTRYPOINT ["python", "run.py"]
 ---> Using cache
 ---> e56614d628d2

Successfully built e56614d628d2
Successfully tagged resources:latest
Creating pg01 ... done
Creating resources01 ... done
Attaching to pg01, resources01
pg01         | 2018-10-25 04:03:03.287 UTC [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
pg01         | 2018-10-25 04:03:03.287 UTC [1] LOG:  listening on IPv6 address "::", port 5432
pg01         | 2018-10-25 04:03:03.294 UTC [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
pg01         | 2018-10-25 04:03:03.317 UTC [20] LOG:  database system was shut down at 2018-10-25 04:02:45 UTC
pg01         | 2018-10-25 04:03:03.321 UTC [1] LOG:  database system is ready to accept connections
resources01 exited with code 0
```
With this you get a exit code of 0 so that the other build steps can proceed. Correct me if I'm wrong about this. Shall we proceed with ` ENTRYPOINT ["python", "run.py"]`?

Reply:
I've resoleved the conflicts in `Dockerfile` to use `ENTRYPOINT ["python", "run.py"]`. Let me know if this works for you.

Reply:
First example:
```Step 12/12 : CMD ["python", "flask", "-m", "run.py"]```
this is definitely not right as the correct order is `python -m $module $module args`

Second example:
```Step 12/12 : CMD ["python", "-m", "flask", "run"]```
this gives the desired behavior as we would sometimes want to run it not in a daemon but want to view the logs as it occurs. 

third example:
``` ENTRYPOINT ["python", "run.py"]```
I don't know the use of a simple `make run`, `make` or `make all` command  that just verifies it can start and then doesn't. 

apparently if one is defined at runtime `CMD` or `ENTRYPOINT` they have the same effect. 

override `ENTRYPOINT`
`docker run --entrypoint [my_entrypoint] test  `
override `CMD`
`docker run test [command 1] [arg1] [arg2]  `

discussion of the [two](http://www.johnzaccone.io/entrypoint-vs-cmd-back-to-basics/) 
```
Use ENTRYPOINT if you don't want developers to change the executable that is run when the container starts. You can think of your container as an "executable wrapper". A good strategy is to define a "stable" combination of executable + parameters as the ENTRYPOINT. Then you can (optionally) specify a default CMD that developers can easily override.
```
it looks like if you do a `CMD` and `ENTRYPOINT`

it has the effect of concatenating the two 
`docker run $container`
`$CMD ENTRYPOINT`

Reply:
So would you prefer defining both `ENTRYPOINT` + `CMD` in the Dockerfile? Also, in `.travis.yml` we may need to change `- sleep 30` to `- sleep 10` as the Travis builds have a timeout of 10 minutes anyhow. If we leave a process open I'm pretty sure that the build will just timeout after running the flask command. We should look at using the matrix build methods if we're going to use the flask command IMO.

Reply:
Nevermind on the `-sleep 30` part. I just examined the logs and it's 30 seconds, not minutes. :zzz:

Reply:
I've added both `ENTRYPOINT ["python", "run.py"]` and `CMD ["python", "-m", "flask", "run"]` in Dockerfile.